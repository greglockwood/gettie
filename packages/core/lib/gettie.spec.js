import Gettie from './gettie';

describe('gettie', () => {
  let originalObj = {
    foo: 123,
    bar: [123, { baz: [1, 2, 3], quux: 4 }],
  };

  it('should be wrapped/unwrapped', () => {
    const gettie = new Gettie(originalObj);
    let obj = gettie.get();

    expect(obj).not.toBe(originalObj);
    expect(gettie.coverage()).toMatchSnapshot();
    expect(gettie.unwrap()).toStrictEqual(originalObj);
  });

  describe('coverage', () => {
    it('just coverage', () => {
      const gettie = new Gettie();
      let obj = gettie.update(originalObj);

      obj.bar[1].quux;
      obj.bar[0];
      expect(gettie.coverage()).toMatchSnapshot();
    });

    it('nullish prop', () => {
      const gettie = new Gettie();
      let obj = gettie.update({ ...originalObj });

      obj.bar[1].quux;
      obj.bar[0];
      obj.bar = null;
      obj.not_defined;
      expect(gettie.coverage()).toMatchSnapshot();
    });
  });

  it('reset', () => {
    const gettie = new Gettie();
    let obj = gettie.update(originalObj);

    obj.bar[1].quux;
    obj.bar[0];
    gettie.reset();
    gettie.update(originalObj);
    expect(gettie.coverage()).toMatchSnapshot();
    obj.bar[1].quux;
    obj.bar[0];
    gettie.reset(originalObj);
    expect(gettie.coverage()).toMatchSnapshot();
  });

  describe('wrap fn', () => {
    it('with lock', () => {
      const gettie = new Gettie();
      const reducer = (state, action) => {
        return { ...state, ...action.payload };
      };
      const newReducer = gettie.wrapFn(reducer);
      const state = newReducer({ foo: 123, bar: 456 }, { payload: { baz: 789 } });
      state.baz;
      newReducer(state, { payload: { quux: 789 } });

      expect(gettie.coverage()).toMatchSnapshot();
    });

    it('without lock', () => {
      const gettie = new Gettie();
      const reducer = (state, action) => {
        return { ...state, ...action.payload };
      };
      const newReducer = gettie.wrapFn(reducer, false);
      const state = newReducer({ foo: 123, bar: 456 }, { payload: { baz: 789 } });
      state.baz;
      newReducer(state, { payload: { quux: 789 } });

      expect(gettie.coverage()).toMatchSnapshot();
    });
  });

  describe('ignoring', () => {
    it('object', () => {
      const gettie = new Gettie();
      let obj = gettie.update({
        foo: 123,
        bar: [123, { [Gettie.ignoreSymbol]: true, baz: [1, 2, 3], quux: 4 }],
      });

      obj.bar[1].quux;
      obj.bar[0];
      expect(gettie.coverage()).toMatchSnapshot();
    });

    it('prop', () => {
      const gettie = new Gettie();
      let obj = gettie.update({
        foo: 123,
        bar: [123, { [Gettie.ignoreSymbol]: ['quux'], baz: [1, 2, 3], quux: 4 }],
      });

      obj.bar[1].baz;
      obj.bar[1].quux;
      obj.bar[0];
      expect(gettie.coverage()).toMatchSnapshot();
    });

    it('symbols', () => {
      const symbol = Symbol('foo');
      const gettie = new Gettie();
      let obj = gettie.update({
        foo: 123,
        [symbol]: { foo: 123 },
      });

      obj.foo;
      obj[symbol].foo;
      expect(gettie.coverage()).toMatchSnapshot();
    });
  });
});
