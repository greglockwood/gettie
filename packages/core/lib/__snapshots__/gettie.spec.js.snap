// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gettie coverage just coverage 1`] = `
Object {
  "branches": Object {
    "all": Array [
      Array [
        "foo",
      ],
      Array [
        "bar",
      ],
      Array [
        "bar",
        0,
      ],
      Array [
        "bar",
        1,
      ],
      Array [
        "bar",
        1,
        "baz",
      ],
      Array [
        "bar",
        1,
        "baz",
        0,
      ],
      Array [
        "bar",
        1,
        "baz",
        1,
      ],
      Array [
        "bar",
        1,
        "baz",
        2,
      ],
      Array [
        "bar",
        1,
        "quux",
      ],
    ],
    "unused": Array [
      Array [
        "foo",
      ],
      Array [
        "bar",
        1,
        "baz",
      ],
      Array [
        "bar",
        1,
        "baz",
        0,
      ],
      Array [
        "bar",
        1,
        "baz",
        1,
      ],
      Array [
        "bar",
        1,
        "baz",
        2,
      ],
    ],
    "used": Array [
      Array [
        "bar",
      ],
      Array [
        "bar",
        "1",
      ],
      Array [
        "bar",
        "1",
        "quux",
      ],
      Array [
        "bar",
        "0",
      ],
    ],
  },
  "data": Object {
    "all": Object {
      "bar": Array [
        123,
        Object {
          "baz": Array [
            1,
            2,
            3,
          ],
          "quux": 4,
        },
      ],
      "foo": 123,
    },
    "unused": Object {
      "bar": Array [
        undefined,
        Object {
          "baz": Array [
            1,
            2,
            3,
          ],
        },
      ],
      "foo": 123,
    },
    "used": Object {
      "bar": Array [
        123,
        Object {
          "quux": 4,
        },
      ],
    },
  },
}
`;

exports[`gettie coverage nullish prop 1`] = `
Object {
  "branches": Object {
    "all": Array [
      Array [
        "foo",
      ],
      Array [
        "bar",
      ],
    ],
    "unused": Array [
      Array [
        "foo",
      ],
    ],
    "used": Array [
      Array [
        "bar",
      ],
      Array [
        "bar",
        "1",
      ],
      Array [
        "bar",
        "1",
        "quux",
      ],
      Array [
        "bar",
        "0",
      ],
      Array [
        "not_defined",
      ],
    ],
  },
  "data": Object {
    "all": Object {
      "bar": null,
      "foo": 123,
    },
    "unused": Object {
      "foo": 123,
    },
    "used": Object {
      "bar": null,
    },
  },
}
`;

exports[`gettie ignoring object 1`] = `
Object {
  "branches": Object {
    "all": Array [
      Array [
        "foo",
      ],
      Array [
        "bar",
      ],
      Array [
        "bar",
        0,
      ],
    ],
    "unused": Array [
      Array [
        "foo",
      ],
    ],
    "used": Array [
      Array [
        "bar",
      ],
      Array [
        "bar",
        "0",
      ],
    ],
  },
  "data": Object {
    "all": Object {
      "bar": Array [
        123,
        Object {
          "baz": Array [
            1,
            2,
            3,
          ],
          "quux": 4,
          Symbol(gettie-ignore): true,
        },
      ],
      "foo": 123,
    },
    "unused": Object {
      "foo": 123,
    },
    "used": Object {
      "bar": Array [
        123,
      ],
    },
  },
}
`;

exports[`gettie ignoring prop 1`] = `
Object {
  "branches": Object {
    "all": Array [
      Array [
        "foo",
      ],
      Array [
        "bar",
      ],
      Array [
        "bar",
        0,
      ],
      Array [
        "bar",
        1,
      ],
      Array [
        "bar",
        1,
        "baz",
      ],
      Array [
        "bar",
        1,
        "baz",
        0,
      ],
      Array [
        "bar",
        1,
        "baz",
        1,
      ],
      Array [
        "bar",
        1,
        "baz",
        2,
      ],
    ],
    "unused": Array [
      Array [
        "foo",
      ],
      Array [
        "bar",
        1,
        "baz",
        0,
      ],
      Array [
        "bar",
        1,
        "baz",
        1,
      ],
      Array [
        "bar",
        1,
        "baz",
        2,
      ],
    ],
    "used": Array [
      Array [
        "bar",
      ],
      Array [
        "bar",
        "1",
      ],
      Array [
        "bar",
        "1",
        "baz",
      ],
      Array [
        "bar",
        "0",
      ],
    ],
  },
  "data": Object {
    "all": Object {
      "bar": Array [
        123,
        Object {
          "baz": Array [
            1,
            2,
            3,
          ],
          "quux": 4,
          Symbol(gettie-ignore): Array [
            "quux",
          ],
        },
      ],
      "foo": 123,
    },
    "unused": Object {
      "bar": Array [
        undefined,
        Object {
          "baz": Array [
            1,
            2,
            3,
          ],
        },
      ],
      "foo": 123,
    },
    "used": Object {
      "bar": Array [
        123,
        Object {
          "baz": Array [],
        },
      ],
    },
  },
}
`;

exports[`gettie ignoring symbols 1`] = `
Object {
  "branches": Object {
    "all": Array [
      Array [
        "foo",
      ],
    ],
    "unused": Array [],
    "used": Array [
      Array [
        "foo",
      ],
    ],
  },
  "data": Object {
    "all": Object {
      "foo": 123,
      Symbol(foo): Object {
        "foo": 123,
      },
    },
    "unused": Object {},
    "used": Object {
      "foo": 123,
    },
  },
}
`;

exports[`gettie reset 1`] = `
Object {
  "branches": Object {
    "all": Array [
      Array [
        "foo",
      ],
      Array [
        "bar",
      ],
      Array [
        "bar",
        0,
      ],
      Array [
        "bar",
        1,
      ],
      Array [
        "bar",
        1,
        "baz",
      ],
      Array [
        "bar",
        1,
        "baz",
        0,
      ],
      Array [
        "bar",
        1,
        "baz",
        1,
      ],
      Array [
        "bar",
        1,
        "baz",
        2,
      ],
      Array [
        "bar",
        1,
        "quux",
      ],
    ],
    "unused": Array [
      Array [
        "foo",
      ],
      Array [
        "bar",
      ],
      Array [
        "bar",
        0,
      ],
      Array [
        "bar",
        1,
      ],
      Array [
        "bar",
        1,
        "baz",
      ],
      Array [
        "bar",
        1,
        "baz",
        0,
      ],
      Array [
        "bar",
        1,
        "baz",
        1,
      ],
      Array [
        "bar",
        1,
        "baz",
        2,
      ],
      Array [
        "bar",
        1,
        "quux",
      ],
    ],
    "used": Array [],
  },
  "data": Object {
    "all": Object {
      "bar": Array [
        123,
        Object {
          "baz": Array [
            1,
            2,
            3,
          ],
          "quux": 4,
        },
      ],
      "foo": 123,
    },
    "unused": Object {
      "bar": Array [
        123,
        Object {
          "baz": Array [
            1,
            2,
            3,
          ],
          "quux": 4,
        },
      ],
      "foo": 123,
    },
    "used": Object {},
  },
}
`;

exports[`gettie reset 2`] = `
Object {
  "branches": Object {
    "all": Array [
      Array [
        "foo",
      ],
      Array [
        "bar",
      ],
      Array [
        "bar",
        0,
      ],
      Array [
        "bar",
        1,
      ],
      Array [
        "bar",
        1,
        "baz",
      ],
      Array [
        "bar",
        1,
        "baz",
        0,
      ],
      Array [
        "bar",
        1,
        "baz",
        1,
      ],
      Array [
        "bar",
        1,
        "baz",
        2,
      ],
      Array [
        "bar",
        1,
        "quux",
      ],
    ],
    "unused": Array [
      Array [
        "foo",
      ],
      Array [
        "bar",
      ],
      Array [
        "bar",
        0,
      ],
      Array [
        "bar",
        1,
      ],
      Array [
        "bar",
        1,
        "baz",
      ],
      Array [
        "bar",
        1,
        "baz",
        0,
      ],
      Array [
        "bar",
        1,
        "baz",
        1,
      ],
      Array [
        "bar",
        1,
        "baz",
        2,
      ],
      Array [
        "bar",
        1,
        "quux",
      ],
    ],
    "used": Array [],
  },
  "data": Object {
    "all": Object {
      "bar": Array [
        123,
        Object {
          "baz": Array [
            1,
            2,
            3,
          ],
          "quux": 4,
        },
      ],
      "foo": 123,
    },
    "unused": Object {
      "bar": Array [
        123,
        Object {
          "baz": Array [
            1,
            2,
            3,
          ],
          "quux": 4,
        },
      ],
      "foo": 123,
    },
    "used": Object {},
  },
}
`;

exports[`gettie should be wrapped/unwrapped 1`] = `
Object {
  "branches": Object {
    "all": Array [
      Array [
        "foo",
      ],
      Array [
        "bar",
      ],
      Array [
        "bar",
        0,
      ],
      Array [
        "bar",
        1,
      ],
      Array [
        "bar",
        1,
        "baz",
      ],
      Array [
        "bar",
        1,
        "baz",
        0,
      ],
      Array [
        "bar",
        1,
        "baz",
        1,
      ],
      Array [
        "bar",
        1,
        "baz",
        2,
      ],
      Array [
        "bar",
        1,
        "quux",
      ],
    ],
    "unused": Array [
      Array [
        "foo",
      ],
      Array [
        "bar",
      ],
      Array [
        "bar",
        0,
      ],
      Array [
        "bar",
        1,
      ],
      Array [
        "bar",
        1,
        "baz",
      ],
      Array [
        "bar",
        1,
        "baz",
        0,
      ],
      Array [
        "bar",
        1,
        "baz",
        1,
      ],
      Array [
        "bar",
        1,
        "baz",
        2,
      ],
      Array [
        "bar",
        1,
        "quux",
      ],
    ],
    "used": Array [],
  },
  "data": Object {
    "all": Object {
      "bar": Array [
        123,
        Object {
          "baz": Array [
            1,
            2,
            3,
          ],
          "quux": 4,
        },
      ],
      "foo": 123,
    },
    "unused": Object {
      "bar": Array [
        123,
        Object {
          "baz": Array [
            1,
            2,
            3,
          ],
          "quux": 4,
        },
      ],
      "foo": 123,
    },
    "used": Object {},
  },
}
`;

exports[`gettie wrap fn with lock 1`] = `
Object {
  "branches": Object {
    "all": Array [
      Array [
        "foo",
      ],
      Array [
        "bar",
      ],
      Array [
        "baz",
      ],
      Array [
        "quux",
      ],
    ],
    "unused": Array [
      Array [
        "foo",
      ],
      Array [
        "bar",
      ],
      Array [
        "quux",
      ],
    ],
    "used": Array [
      Array [
        "baz",
      ],
    ],
  },
  "data": Object {
    "all": Object {
      "bar": 456,
      "baz": 789,
      "foo": 123,
      "quux": 789,
    },
    "unused": Object {
      "bar": 456,
      "foo": 123,
      "quux": 789,
    },
    "used": Object {
      "baz": 789,
    },
  },
}
`;

exports[`gettie wrap fn without lock 1`] = `
Object {
  "branches": Object {
    "all": Array [
      Array [
        "foo",
      ],
      Array [
        "bar",
      ],
      Array [
        "baz",
      ],
      Array [
        "quux",
      ],
    ],
    "unused": Array [
      Array [
        "quux",
      ],
    ],
    "used": Array [
      Array [
        "baz",
      ],
      Array [
        "foo",
      ],
      Array [
        "bar",
      ],
    ],
  },
  "data": Object {
    "all": Object {
      "bar": 456,
      "baz": 789,
      "foo": 123,
      "quux": 789,
    },
    "unused": Object {
      "quux": 789,
    },
    "used": Object {
      "bar": 456,
      "baz": 789,
      "foo": 123,
    },
  },
}
`;
